blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+| repository                    | commit              | upload_id | distance |
+| ----------------------------- | ------------------- | --------- | -------- |
+| github.com/sourcegraph/sample | <code>323e23</code> | 1         | 1        |
+| github.com/sourcegraph/sample | <code>d67b8d</code> | 1         | 0        |
+| github.com/sourcegraph/sample | <code>4c8d9d</code> | 1         | 1        |
+| github.com/sourcegraph/sample | <code>4c8d9d</code> | 2         | 1        |
+| github.com/sourcegraph/sample | <code>f4fb06</code> | 2         | 0        |
+| github.com/sourcegraph/sample | <code>a36064</code> | 2         | 1        |
+| github.com/sourcegraph/sample | <code>313082</code> | 1         | 2        |
+| github.com/sourcegraph/sample | <code>6a06fc</code> | 1         | 3        |
+
+The `lsif_dirty_repositories` table tracks which repositories need their commit graphs updated. When we receive an upload for a repository, or get a request for a commit that we don't currently track, we bump the `dirty_token` value attached to that repository. When we are about to refresh the graph, we note the dirty token, calculate the set of visible uploads for each commit, write it to the database, and set the `update_token` to the value of the dirty token we noted earlier. This ensures that we avoid a particular class of race conditions that occur when we receive an upload at the same time we're re-calculating the commit graph from a previous upload.
+
+| repository                    | dirty_token | update_token |
+| ----------------------------- | ----------- | ------------ |
+| github.com/sourcegraph/sample | 42          | 42           |
+
+#### Example
+
+For this example, we'll use the same graph as we did above, where commits `80c800`, `c85b4b`, and `3daedb` define uploads #1, #2, and #3, respectively.
+
+<div class="no-shadow">
+  <img src="https://sourcegraphstatic.com/blog/commit-graph-optimizations/graph3.png" alt="sample commit graph">
+</div>
+
+The first step to the algorithm is to [topologically order](https://en.wikipedia.org/wiki/Topological_sorting) each commit so that a commit is processed only after all of its parents are processed. Then we perform a simple [greedy process](https://en.wikipedia.org/wiki/Greedy_algorithm) to determine which uploads are visible at each stage of the commit. If a commit defines an upload, it is visible to that commit. Otherwise, the set of visible uploads for the commit is the same as the uploads visible from each parent (just at a higher distance).
```

Comment:
```suggestion
The first step to the algorithm is to [topologically sort](https://en.wikipedia.org/wiki/Topological_sorting) each commit so that a commit is processed only after all of its parents are processed. Then we perform a simple [greedy process](https://en.wikipedia.org/wiki/Greedy_algorithm) to determine which uploads are visible at each stage of the commit. If a commit defines an upload, it is visible to that commit. Otherwise, the set of visible uploads for the commit is the same as the uploads visible from each parent (just at a higher distance).
```